specs/airr-schema.yaml

```diff
@@ -2259,118 +2253,152 @@ Clone:
             type: string
             description: >
                 Assembled, aligned, full-length inferred germline sequence spanning the same region
                 as the sequence_alignment field (typically the V(D)J region) and including the same set
                 of corrections and spacers (if any).
+            x-airr:
+                miairr: false
         germline_alignment_aa:
             type: string
             description: >
                 Amino acid translation of the assembled germline sequence.
+            x-airr:
+                miairr: false
         v_alignment_start:
             type: integer
             description: >
                 Start position in the V segment in both the sequence_alignment and germline_alignment fields
                 (1-based closed interval).
+            x-airr:
+                miairr: false
+                required: true
         v_alignment_end:
             type: integer
             description: >
                 End position in the V segment in both the sequence_alignment and germline_alignment fields
                 (1-based closed interval).
+            x-airr:
+                miairr: false
+                required: true
         d_alignment_start:
             type: integer
             description: >
                 Start position of the D segment in both the sequence_alignment and germline_alignment fields
                 (1-based closed interval).
+            x-airr:
+                miairr: false
+                required: true
         d_alignment_end:
             type: integer
             description: >
                 End position of the D segment in both the sequence_alignment and germline_alignment fields
                 (1-based closed interval).
+            x-airr:
+                miairr: false
+                required: true
         j_alignment_start:
             type: integer
             description: >
                 Start position of the J segment in both the sequence_alignment and germline_alignment fields
                 (1-based closed interval).
+            x-airr:
+                miairr: false
+                required: true
         j_alignment_end:
             type: integer
             description: >
                 End position of the J segment in both the sequence_alignment and germline_alignment fields
                 (1-based closed interval).
+            x-airr:
+                miairr: false
+                required: true
         cdr3_start:
             type: integer
             description: CDR3 start position in the query sequence (1-based closed interval).
+            x-airr:
+                miairr: false
+                required: true
         cdr3_end:
             type: integer
             description: CDR3 end position in the query sequence (1-based closed interval).
+            x-airr:
+                miairr: false
+                required: true
         rearrangement_count:
```

Comment:
Do we want a list of `sequence_id`s from the original `Repertoire` that are included in this clone?

Reply:
IIRC, those are supposed to be in each `Node`. I pushed a change replacing `node_id` with `sequence_id`, moving the junction sequence fields to Node as well.

But... that's only going to make sense if you have a Tree object. Otherwise, you'll have to go back to the `clone_id` column in the Rearrangement data to get the list of `sequence_id` values. Which might be fine?

Reply:
`node_id` is not necessarily equivalent to `sequence_id`, because of internal nodes. In addition, while we could require that (for leaves) `node_id` has to match a `sequence_id` in `Repertoire`, I don't think that's currently documented as part of the schema.

Finally, while this is mostly relevant to olmstead, at the moment, if you build trees from a subset of the sequences in a clone, than even accessing the/a `Tree` object and collecting all of the `Nodes` doesn't guarantee that you will catch all of the members.

Reply:
All of the annotation properties should belong to the `Clone` either because they by definition refer to the whole clone (`v_call` etc) or because they refer to the `naive_ancestor` (`junction_start` etc).
If the clone contains sequences with indels relative to the inferred `naive_ancestor`, those should (for the moment) probably be added as `Node`-level properties. For alignments etc of individual sequences, a user should either cross-reference the original `Rearrangement` or compare to the `naive_ancestor` using the indel information present on the `Node`. I don't think we want/expect this kind of mostly redundant information to be re-encoded at every `Node`.

Reply:
> node_id is not necessarily equivalent to sequence_id, because of internal nodes. In addition, while we could require that (for leaves) node_id has to match a sequence_id in Repertoire, I don't think that's currently documented as part of the schema.

I think it's fine to give internal nodes their own unique `sequence_id`, because `sequence_id` is already pretty context dependent in terms of what it means. There's also a use case where you'd want to dump inferred nodes out to a Rearrangement TSV to analyze them as you would any other sequence (particularly the MRCA), so you wouldn't have to do any translation of column names if we just stick with `sequence_id`.

> All of the annotation properties should belong to the Clone either because they by definition refer to the whole clone (v_call etc) or because they refer to the naive_ancestor (junction_start etc).

I was kind of scratching my head on this one too. I'm not sure whether it's a problem. How it's setup now it looks like there's an assumption that all members of a clone have the same junction length. That's a common approach by tools, but not universal. But... everything is supposed to be operating on a multiple alignment of the sequences/germline, so indels are handled by that requirement.

Maybe the solution is to not make some of the fields required in `Clone` and copy them into `Node` as well. So that you can chose whether to defined certain field across all Nodes in a clone or on a Node-by-Node basis.

All those positional fields shouldn't be required though, if they aren't essential to the tree building process and aren't required in the Rearrangement schema.

Reply:
**node_id:**
> this is mostly relevant to olmstead, at the moment, if you build trees from a subset of the sequences in a clone, than even accessing the/a Tree object and collecting all of the Nodes doesn't guarantee that you will catch all of the members.

In Olmsted we don't currently keep track of a clone's sequences that we didn't include in a tree for that clone except summary statistics at the clone level (e.g. how many sequences belong to the clone). For our use case I'm ok with either
>  give internal nodes their own unique sequence_id, because sequence_id

, or having `node_id` be separate.


**annotation properties:**
I think it makes sense to keep these at the clone level as @scharch says. It seems 
> everything is supposed to be operating on a multiple alignment of the sequences/germline, so indels are handled by that requirement.

means that this model covers the case the @javh is pointing out, if I understand correctly.

> All those positional fields shouldn't be required though, if they aren't essential to the tree building process and aren't required in the Rearrangement schema.

I think I made them required since some of them are required by or are products of the clone inference process in partis. If this is not broadly applicable, please remove the required property from any of them.

Reply:
My summary of this thread:
* `Clone` should have an array of `clonal_sequence`s (?) that match `sequence_id`s in the `Repertoire`.
* `germline_alignment` should be a required field for `Clone`. I'd still like the description to be tweaked slightly to point out that it the inferred ancestor for the entire `Clone` and that the relevant `sequence_alignment`s are those of the `clonal_sequence`s (as opposed to in the Rearrangements TSV, where both fields are at the same level).
* `v_call` etc should be fields at the `Clone` level but not required
* `node_id` should be defined to match `sequence_id` in the source `Repertoire` where relevant/possible. (NB this could get dicey if analyzing multiple `Repertoire`s together.)

@javh @eharkins look ok?

Reply:
1. I don't think this is necessary, because the list of sequences in a clone will be contained in the Node objects. You can also get the `sequence_id` in the Clone from the Rearrangement object by just keying on `clone_id`. That said, having a non-required field that contains a list of `sequence_id`s seems fine to me. I would name it `sequence_id` though, because the semantics are the same as `sequence_id` in the Rearrangement object - it's just a list instead of a scalar.
2. Agreed. I think tweaking the description is a good idea. That's simpler than making a new field.
3. Agreed.
4. Yes, but I would just use `sequence_id` instead of making a new `node_id` field that contains largely the same information.

Reply:
> 1. I don't think this is necessary, because the list of sequences in a clone will be contained in the Node objects.

I disagree, because there is not guarantee that all clonal members are used to build the tree (which could be subsampled) and therefore no guarantee that they will be represented by a `Node` object. However, I guess I am ok with making it a non-required field.

Totally fine with the nomenclature refinements you suggest @javh.

Reply:
@scharch, okay, then I agree we should include the `sequence_id` list in Clone then. This is for testing purposes after all, so we can always tweak it later.

Reply:
This sounds good to me. @scharch do you want to push a commit with the changes you listed above with @javh's amendments:
> 1 ... having a non-required field that contains a list of sequence_ids seems fine to me. I would name it sequence_id though, because the semantics are the same as sequence_id in the Rearrangement object - it's just a list instead of a scalar.

> 4. Yes, but I would just use sequence_id instead of making a new node_id field that contains largely the same information.

if we agree on them?